src/Orleans/Threading/CancellationSourcesExtension.cs

```diff
@@ -0,0 +1,52 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Orleans.Concurrency;
+using Orleans.Runtime;
+
+namespace Orleans.Threading
+{
+    internal class CancellationSourcesExtension : ICancellationSourcesExtension
+    {
+        private readonly Lazy<TraceLogger> _logger = new Lazy<TraceLogger>(() =>
+            TraceLogger.GetLogger("CancellationSourcesExtension", TraceLogger.LoggerType.Application));
+        private readonly Dictionary<Guid, CancellationTokenSource> _cancellationTokenSources;
```

Comment:
The actuall memory consumption from dictionary with 1 million guid - cts pairs was 100 megabytes, so, assuming that multiple `AsyncTaskSafeTimer` would put additional load on the scheduler and the in memory tokens are being cleaned on grain deactivation there's probably no need to clean old tokens. An alternative is to use weak references and run timer after some threshold exceeding.